File: educational-core/src/com/jetbrains/edu/learning/EduState.java
package com.jetbrains.edu.learning;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import com.jetbrains.edu.learning.editor.EduEditor;
import org.jetbrains.annotations.Nullable;

public class EduState {
  private final EduEditor myEduEditor;
  private final Editor myEditor;
  private final TaskFile myTaskFile;
  private final VirtualFile myVirtualFile;
  private final Task myTask;

  public EduState(@Nullable final EduEditor eduEditor) {
    myEduEditor = eduEditor;
    myEditor = eduEditor != null ? eduEditor.getEditor() : null;
    myTaskFile = eduEditor != null ? eduEditor.getTaskFile() : null;
    myVirtualFile = myEditor != null ? FileDocumentManager.getInstance().getFile(myEditor.getDocument()) : null;
    myTask = myTaskFile != null ? myTaskFile.getTask() : null;
  }

  public Editor getEditor() {
    return myEditor;
  }

  public TaskFile getTaskFile() {
    return myTaskFile;
  }

  public VirtualFile getVirtualFile() {
    return myVirtualFile;
  }

  public Task getTask() {
    return myTask;
  }

  public boolean isValid() {
    return myEduEditor != null && myEditor != null &&
           myTaskFile != null && myVirtualFile != null &&
           myTask != null;
  }
}

File: educational-core/src/com/jetbrains/edu/learning/EduState.kt
package com.jetbrains.edu.learning

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.jetbrains.edu.learning.courseFormat.TaskFile
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.editor.EduEditor


class EduState private constructor(eduEditor: EduEditor, val virtualFile: VirtualFile) {
  val editor: Editor = eduEditor.editor
  val taskFile: TaskFile = eduEditor.taskFile
  val task: Task = taskFile.task

  companion object {
    @JvmStatic
    fun getEduState(project: Project): EduState? {
      val eduEditor = EduUtils.getSelectedEduEditor(project) ?: return null
      val virtualFile = FileDocumentManager.getInstance().getFile(eduEditor.editor.document) ?: return null
      return EduState(eduEditor, virtualFile)
    }
  }
}